### PR TITLE
[server]  integration test for blob transfer between servers

### PR DESCRIPTION
**Pull request by jingy-li**
## Integration test for blob transfer between servers

- Minor bug fix for ServerBlobFinder. Previously initialize ServerBlobFinder need to wait for customizedViewRepository complete, it waits so long. Moving get customizedViewRepository when we need to find peer not in the constructor. 
- Add integration test to do 
    - Start 2 servers and a VPJ. 
    - Remove one servers metadata partition and data partition 
    - Restart server1. 
    - Verify that the re-started server successfully ingest data from server2 via blob transfer. 

## How was this PR tested?
(1) Integration test: the partition file is transferred and the metadata is transferred and updated. 
(2) Adding logs in updateStorePartitionMetadata to compare the offset record before and after the metadata update. Offset record is successfully transferred and used.  

Screenshot 2024-09-27 at 5 49 12 PM
Screenshot 2024-09-27 at 5 49 46 PM

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+      OffsetRecord offsetServer1 =
+          server1.getVeniceServer().getStorageMetadataService().getLastOffset("test-store_v1", i);
+      OffsetRecord offsetServer2 =
+          server2.getVeniceServer().getStorageMetadataService().getLastOffset("test-store_v1", i);
+      Assert.assertEquals(offsetServer1.getLocalVersionTopicOffset(), offsetServer2.getLocalVersionTopicOffset());
+    }
+  }
+
+  public VeniceClusterWrapper initializeVeniceCluster() {
```
Comment (adamxchen) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:84]: The practice we usually do it to move codes related to setup env to a method, say `setup`, then add testNG annotation to let it run before test cases so we can potentially add more test cases for this suite. Can you plz do that?

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    // cleanup and restart server 1
+    FileUtils.deleteDirectory(
+        new File(RocksDBUtils.composePartitionDbDir(path1 + "/rocksdb", storeName + "_v1", METADATA_PARTITION_ID)));
+    FileUtils.deleteDirectory(new File(RocksDBUtils.composePartitionDbDir(path1 + "/rocksdb", storeName + "_v1", 0)));
+    FileUtils.deleteDirectory(new File(RocksDBUtils.composePartitionDbDir(path1 + "/rocksdb", storeName + "_v1", 1)));
+    FileUtils.deleteDirectory(new File(RocksDBUtils.composePartitionDbDir(path1 + "/rocksdb", storeName + "_v1", 2)));
+    cluster.stopAndRestartVeniceServer(server1.getPort());
+
+    // wait 2 min
+    Thread.sleep(120000);
```
Comment (adamxchen) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:68]: can we use `TestUtils.waitForNonDeterministicAssertion`?

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    // the partition files should be transferred to server 1 and offset should be the same
```
Comment (adamxchen) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:70]: The concern is we don't know whether the server ingests the data via blob transfer or the kafka.  Yes, the logging can help us here but it's not a programmatic way and the chance is someone breaks the feature in the future and we don't know.

We discussed this in [another thread](https://github.com/linkedin/venice/pull/1118#discussion_r1719109319) a while ago. Do you think we can leverage that to validate?

Comment (adamxchen) [reply internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:83]: discussed offline. The IDENTITY file changes so it cannot be used to verify whether partitions really happened or not. There's no reason to block this PR at least and we will come back to the topic later on.

**Review by adamxchen — APPROVED** [review]
Thanks!

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+  public void testBlobP2PTransferAmongServers() throws Exception {
+    String storeName = "test-store";
+    Consumer<UpdateStoreQueryParams> paramsConsumer = params -> params.setBlobTransferEnabled(true);
+    setUpStore(cluster, storeName, paramsConsumer, properties -> {}, true);
+
+    VeniceServerWrapper server1 = cluster.getVeniceServers().get(0);
+    VeniceServerWrapper server2 = cluster.getVeniceServers().get(1);
+
+    // very the snapshot is generated for both servers after the job is done
+    for (int i = 0; i < 3; i++) {
```
Comment (lluwm) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:64]: Nit, would it be more readable if we have a descriptively named constants here?

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+      String snapshotPath2 = RocksDBUtils.composeSnapshotDir(path2 + "/rocksdb", storeName + "_v1", i);
+      Assert.assertTrue(Files.exists(Paths.get(snapshotPath2)));
+    }
+
+    // cleanup and restart server 1
+    FileUtils.deleteDirectory(
+        new File(RocksDBUtils.composePartitionDbDir(path1 + "/rocksdb", storeName + "_v1", METADATA_PARTITION_ID)));
+    FileUtils.deleteDirectory(new File(RocksDBUtils.composePartitionDbDir(path1 + "/rocksdb", storeName + "_v1", 0)));
+    FileUtils.deleteDirectory(new File(RocksDBUtils.composePartitionDbDir(path1 + "/rocksdb", storeName + "_v1", 1)));
+    FileUtils.deleteDirectory(new File(RocksDBUtils.composePartitionDbDir(path1 + "/rocksdb", storeName + "_v1", 2)));
```
Comment (lluwm) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:76]: Nit, seems we can put it into a small loop for simplicity? Also, can we add a step to verify that files are not existent after the delete?

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+  @Test(singleThreaded = true)
+  public void testBlobP2PTransferAmongServers() throws Exception {
+    String storeName = "test-store";
+    Consumer<UpdateStoreQueryParams> paramsConsumer = params -> params.setBlobTransferEnabled(true);
+    setUpStore(cluster, storeName, paramsConsumer, properties -> {}, true);
+
+    VeniceServerWrapper server1 = cluster.getVeniceServers().get(0);
+    VeniceServerWrapper server2 = cluster.getVeniceServers().get(1);
+
+    // very the snapshot is generated for both servers after the job is done
```
Comment (lluwm) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:63]: Typo: very-> verify?

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    for (int i = 0; i < 3; i++) {
+      String snapshotPath1 = RocksDBUtils.composeSnapshotDir(path1 + "/rocksdb", storeName + "_v1", i);
+      Assert.assertTrue(Files.exists(Paths.get(snapshotPath1)));
+      String snapshotPath2 = RocksDBUtils.composeSnapshotDir(path2 + "/rocksdb", storeName + "_v1", i);
+      Assert.assertTrue(Files.exists(Paths.get(snapshotPath2)));
+    }
```
Comment (lluwm) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:69]: A dumb question, do we remove snapshot path after the blob transfer is done in the design?

Comment (adamxchen) [reply internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:69]: I missed this. Yes, @jingy-li please double-check and see if we need to do more clean-up here

Comment (jingy-li) [reply internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/BlobP2PTransferAmongServersTest.java:69]: In server side, we do not clean the snapshot when the transfer is done.

**Review by lluwm — APPROVED** [review]
Thank you for this change @jingy-li, LGTM!

**Pull request merged**